File: src/main/java/july_2025/FindDuplicateNumber.java
package july_2025;

public class FindDuplicateNumber {

    public static void main(String[] args) {
        int[] nums = {1,3,4,2,2};
        FindDuplicateNumber findDuplicateNumber = new FindDuplicateNumber();
        int result = findDuplicateNumber.findDuplicate(nums);
        System.out.println(result);
    }

    // O(n) time | O(1) space
    public int findDuplicate(int[] nums) {
        int slow = 0;
        int fast = 0;

        while (true) {
            slow = nums[slow];
            fast = nums[nums[fast]];
            if (slow == fast) {
                break;
            }
        }
        int slow2 = 0;
        while (true) {
            slow = nums[slow];
            slow2 = nums[slow2];
            if (slow == slow2) {
                break;
            }
        }
        return slow2;
    }

}

File: python/second/find_duplicate_number.py
class Solution(object):
    # O(n) time | O(1) space
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        slow = 0
        fast = 0

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
        
        slow2 = 0
        while True:
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow == slow2:
                break
        return slow

solution = Solution()
nums = [1,3,4,2,2]
print(solution.findDuplicate(nums))